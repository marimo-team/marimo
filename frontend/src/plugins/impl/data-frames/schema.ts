/* Copyright 2024 Marimo. All rights reserved. */
import {
  AGGREGATION_FNS,
  NUMPY_DTYPES,
} from "@/plugins/impl/data-frames/types";
import {
  FieldOptions,
  randomNumber,
} from "@/plugins/impl/data-frames/forms/options";
import { z } from "zod";
import {
  ALL_OPERATORS,
  OperatorType,
  isConditionValueValid,
} from "./utils/operators";

const column_id = z
  .string()
  .min(1, "Required")
  .describe(FieldOptions.of({ label: "Column", special: "column_id" }));

const column_id_array = z
  .array(column_id.describe(FieldOptions.of({ special: "column_id" })))
  .min(1, "At least one column is required")
  .default([])
  .describe(FieldOptions.of({ label: "Columns" }));

const ColumnConversionTransformSchema = z
  .object({
    type: z.literal("column_conversion"),
    column_id: column_id,
    data_type: z
      .enum(NUMPY_DTYPES)
      .describe(FieldOptions.of({ label: "Data type (numpy)" }))
      .default("bool"),
    errors: z
      .enum(["ignore", "raise"])
      .default("ignore")
      .describe(
        FieldOptions.of({ label: "Handle errors", special: "radio_group" }),
      ),
  })
  .describe(FieldOptions.of({}));

const RenameColumnTransformSchema = z.object({
  type: z.literal("rename_column"),
  column_id: column_id,
  new_column_id: z
    .string()
    .min(1, "Required")
    .describe(FieldOptions.of({ label: "New column name" })),
});

const SortColumnTransformSchema = z.object({
  type: z.literal("sort_column"),
  column_id: column_id,
  ascending: z
    .boolean()
    .describe(FieldOptions.of({ label: "Ascending" }))
    .default(true),
  na_position: z
    .enum(["first", "last"])
    .describe(
      FieldOptions.of({ label: "N/A position", special: "radio_group" }),
    )
    .default("last"),
});

export const ConditionSchema = z
  .object({
    column_id: column_id,
    operator: z
      .enum(Object.keys(ALL_OPERATORS) as [OperatorType, ...OperatorType[]])
      .describe(FieldOptions.of({ label: " " })),
    value: z.any().describe(FieldOptions.of({ label: "Value" })),
  })
  .describe(FieldOptions.of({ direction: "row", special: "column_filter" }));

const FilterRowsTransformSchema = z.object({
  type: z.literal("filter_rows"),
  operation: z
    .enum(["keep_rows", "remove_rows"])
    .default("keep_rows")
    .describe(FieldOptions.of({ special: "radio_group" })),
  where: z
    .array(ConditionSchema)
    .min(1)
    .transform((value) => {
      return value.filter((condition) => {
        return isConditionValueValid(condition.operator, condition.value);
      });
    })
    .describe(FieldOptions.of({ label: "Value" }))
    .default([{ column_id: "", operator: "==", value: "" }]),
});

const GroupByTransformSchema = z
  .object({
    type: z.literal("group_by"),
    column_ids: column_id_array,
    aggregation: z
      .enum(AGGREGATION_FNS)
      .default("count")
      .describe(FieldOptions.of({ label: "Aggregation" })),
    drop_na: z
      .boolean()
      .default(false)
      .describe(FieldOptions.of({ label: "Drop N/A" })),
  })
  .describe(FieldOptions.of({}));

const AggregateTransformSchema = z
  .object({
    type: z.literal("aggregate"),
    column_ids: column_id_array,
    aggregations: z
      .array(z.enum(AGGREGATION_FNS))
      .min(1, "At least one aggregation is required")
      .default(["count"])
      .describe(FieldOptions.of({ label: "Aggregations" })),
  })
  .describe(FieldOptions.of({ direction: "row" }));

const SelectColumnsTransformSchema = z.object({
  type: z.literal("select_columns"),
  column_ids: column_id_array,
});

const SampleRowsTransformSchema = z.object({
  type: z.literal("sample_rows"),
  n: z
    .number()
    .positive()
    .describe(FieldOptions.of({ label: "Number of rows" })),
  seed: z
    .number()
    .default(() => randomNumber())
    .describe(
      FieldOptions.of({ label: "Re-sample", special: "random_number_button" }),
    ),
  replace: z
    .boolean()
    .default(false)
    .describe(
      FieldOptions.of({
        label: "Sample with replacement",
      }),
    ),
});

const ShuffleRowsTransformSchema = z.object({
  type: z.literal("shuffle_rows"),
  seed: z
    .number()
    .default(() => randomNumber())
    .describe(
      FieldOptions.of({ label: "Re-shuffle", special: "random_number_button" }),
    ),
});

export const TransformTypeSchema = z.union([
  FilterRowsTransformSchema,
  SelectColumnsTransformSchema,
  RenameColumnTransformSchema,
  ColumnConversionTransformSchema,
  SortColumnTransformSchema,
  GroupByTransformSchema,
  AggregateTransformSchema,
  SampleRowsTransformSchema,
  ShuffleRowsTransformSchema,
]);

export type TransformType = z.infer<typeof TransformTypeSchema>;

export const TransformationsSchema = z.object({
  transforms: z.array(TransformTypeSchema),
});

export type Transformations = z.infer<typeof TransformationsSchema>;
