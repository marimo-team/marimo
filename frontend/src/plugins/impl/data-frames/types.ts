/* Copyright 2024 Marimo. All rights reserved. */

/**
 * Map of column id to data type
 */
export interface ColumnDataTypes {
  [column_id: string]: string;
}

export const NUMPY_DTYPES = [
  "int8",
  "int16",
  "int32",
  "int64",
  "uint8",
  "uint16",
  "uint32",
  "uint64",
  "float16",
  "float32",
  "float64",
  "float128",
  "complex64",
  "complex128",
  "complex256",
  "bool",
  "object",
  "string_",
  "unicode_",
  "datetime64",
  "timedelta64",
] as const;

export const AGGREGATION_FNS = [
  "count",
  "sum",
  "mean",
  "median",
  "min",
  "max",
] as const;
