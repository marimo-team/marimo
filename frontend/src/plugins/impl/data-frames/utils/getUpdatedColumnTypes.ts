/* Copyright 2024 Marimo. All rights reserved. */
import { logNever } from "@/utils/assertNever";
import { TransformType } from "../schema";
import { Objects } from "@/utils/objects";
import { ColumnDataTypes } from "../types";

/**
 * Given a list of transforms, return the updated column names/types.
 */
export function getUpdatedColumnTypes(
  transforms: TransformType[],
  columnTypes: ColumnDataTypes,
): ColumnDataTypes {
  if (!transforms || transforms.length === 0) {
    return columnTypes;
  }

  let next: ColumnDataTypes = { ...columnTypes };
  for (const transform of transforms) {
    next = handleTransform(transform, next);
  }

  return next;
}

function handleTransform(
  transform: TransformType,
  next: ColumnDataTypes,
): ColumnDataTypes {
  switch (transform.type) {
    case "column_conversion":
      if (!transform.column_id) {
        return next;
      }

      next[transform.column_id] = transform.data_type;
      return next;
    case "rename_column":
      if (!transform.new_column_id) {
        return next;
      }
      return Objects.fromEntries(
        Objects.entries(next).map(([k, v]) => {
          if (k === transform.column_id) {
            return [transform.new_column_id, v];
          }
          return [k, v];
        }),
      );
    case "group_by":
      return Objects.filter(
        next,
        (_v, k) => !transform.column_ids.includes(k as string),
      );
    case "aggregate":
      return Objects.filter(next, (_v, k) =>
        transform.column_ids.includes(k as string),
      );
    case "select_columns":
      return Objects.filter(next, (_v, k) =>
        transform.column_ids.includes(k as string),
      );
    case "filter_rows":
    case "shuffle_rows":
    case "sample_rows":
    case "sort_column":
      return next;
    default:
      logNever(transform);
      return next;
  }
}
