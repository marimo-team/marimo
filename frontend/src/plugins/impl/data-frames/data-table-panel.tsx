/* Copyright 2023 Marimo. All rights reserved. */
import { generateColumns } from "@/components/data-table/columns";
import { DataTable } from "@/components/data-table/data-table";
import { Functions } from "@/utils/functions";
import React, { useMemo } from "react";
import { T } from "vitest/dist/reporters-5f784f42";

interface Props {
  data: T[];
}

export const DataTablePanel: React.FC<Props> = ({ data }) => {
  const columns = useMemo(() => generateColumns(data, null), [data]);

  if (!data) {
    return null;
  }

  return (
    <DataTable
      data={data}
      columns={columns}
      pagination={true}
      selection={null}
      rowSelection={{}}
      onRowSelectionChange={Functions.NOOP}
    />
  );
};
