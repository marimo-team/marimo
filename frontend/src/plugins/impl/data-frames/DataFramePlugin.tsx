/* Copyright 2024 Marimo. All rights reserved. */
import { z } from "zod";

import { Transformations } from "./schema";
import { TransformPanel } from "./panel";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
import { Code2Icon, FunctionSquareIcon } from "lucide-react";
import { CodePanel } from "./python/code-panel";
import { ColumnDataTypes } from "./types";
import { createPlugin } from "@/plugins/core/builder";
import { rpc } from "@/plugins/core/rpc";
import { useAsyncData } from "@/hooks/useAsyncData";
import { LoadingDataTableComponent } from "../DataTablePlugin";
import { Functions } from "@/utils/functions";
import { Arrays } from "@/utils/arrays";
import { useState } from "react";
import { TooltipProvider } from "@/components/ui/tooltip";
import { Banner, ErrorBanner } from "../common/error-banner";

/**
 * Arguments for a data table
 *
 * @param label - a label of the table
 * @param data - the data to display
 */
interface Data {
  label?: string | null;
  columns: ColumnDataTypes;
  dataframeName: string;
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type PluginFunctions = {
  get_dataframe: (req: {}) => Promise<{
    url: string;
    has_more: boolean;
    total_rows: number;
    row_headers: Array<[string, string[]]>;
  }>;
  get_column_values: (req: { column: string }) => Promise<{
    values: unknown[];
    too_many_values: boolean;
  }>;
};

// Value is selection, but it is not currently exposed to the user
type S = Transformations | undefined;

export const DataFramePlugin = createPlugin<S>("marimo-dataframe")
  .withData(
    z.object({
      label: z.string().nullish(),
      dataframeName: z.string(),
      columns: z
        .object({})
        .passthrough()
        .transform((value) => value as ColumnDataTypes),
    }),
  )
  .withFunctions<PluginFunctions>({
    // Get the data as a URL
    get_dataframe: rpc.input(z.object({})).output(
      z.object({
        url: z.string(),
        has_more: z.boolean(),
        total_rows: z.number(),
        row_headers: z.array(z.tuple([z.string(), z.array(z.any())])),
      }),
    ),
    get_column_values: rpc.input(z.object({ column: z.string() })).output(
      z.object({
        values: z.array(z.any()),
        too_many_values: z.boolean(),
      }),
    ),
  })
  .renderer((props) => (
    <TooltipProvider>
      <DataFrameComponent
        {...props.data}
        {...props.functions}
        value={props.value}
        setValue={props.setValue}
      />
    </TooltipProvider>
  ));

interface DataTableProps extends Data, PluginFunctions {
  value: S;
  setValue: (value: S) => void;
}

const EMPTY: Transformations = {
  transforms: [],
};

export const DataFrameComponent = ({
  columns,
  dataframeName,
  value,
  setValue,
  get_dataframe,
  get_column_values,
}: DataTableProps): JSX.Element => {
  const { data, error } = useAsyncData(
    () => get_dataframe({}),
    [value?.transforms],
  );
  const { url, has_more, total_rows, row_headers } = data || {};

  const [internalValue, setInternalValue] = useState<Transformations>(
    value || EMPTY,
  );

  return (
    <div>
      <Tabs defaultValue="transform">
        <TabsList className="h-8">
          <TabsTrigger value="transform" className="text-xs py-1">
            <FunctionSquareIcon className="w-3 h-3 mr-2" />
            Transform
          </TabsTrigger>
          <TabsTrigger value="code" className="text-xs py-1">
            <Code2Icon className="w-3 h-3 mr-2" />
            Code
          </TabsTrigger>
        </TabsList>
        <TabsContent
          value="transform"
          className="mt-1 border rounded-t overflow-hidden"
        >
          <TransformPanel
            initialValue={internalValue}
            columns={columns}
            onChange={(v) => {
              // Update the value valid changes
              setValue(v);
              setInternalValue(v);
            }}
            onInvalidChange={setInternalValue}
            getColumnValues={get_column_values}
          />
        </TabsContent>
        <TabsContent
          value="code"
          className="mt-1 border rounded-t overflow-hidden"
        >
          <CodePanel dataframeName={dataframeName} transforms={value} />
        </TabsContent>
      </Tabs>
      {error && <ErrorBanner error={error} />}
      {has_more && total_rows && (
        <Banner>Result clipped. Total rows {prettyNumber(total_rows)}.</Banner>
      )}
      <LoadingDataTableComponent
        label={null}
        className="rounded-b border-x border-b"
        data={url || ""}
        pageSize={5}
        pagination={true}
        rowHeaders={row_headers || Arrays.EMPTY}
        showDownload={false}
        download_as={Functions.THROW}
        value={Arrays.EMPTY}
        setValue={Functions.NOOP}
        selection={null}
      />
    </div>
  );
};

function prettyNumber(value: number): string {
  return new Intl.NumberFormat().format(value);
}
