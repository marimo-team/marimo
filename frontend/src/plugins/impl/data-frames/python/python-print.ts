/* Copyright 2024 Marimo. All rights reserved. */
import { TransformType } from "@/plugins/impl/data-frames/schema";
import { logNever } from "@/utils/assertNever";
import { OperatorType } from "../utils/operators";

export function pythonPrintTransforms(
  dfName: string,
  transforms: TransformType[],
): string {
  const dfNextName = `${dfName}_next`;
  const transformStrs = transforms.map(
    (transform) => `${dfNextName} = ${pythonPrint(dfNextName, transform)}`,
  );

  return [`${dfNextName} = ${dfName}`, ...transformStrs].join("\n");
}

export function pythonPrint(dfName: string, transform: TransformType): string {
  switch (transform.type) {
    case "column_conversion": {
      const { column_id, data_type, errors } = transform;
      return `${dfName}["${column_id}"].astype("${data_type}", errors="${errors}")`;
    }
    case "rename_column": {
      const { column_id, new_column_id } = transform;
      return `${dfName}.rename(columns={"${column_id}": "${new_column_id}"})`;
    }
    case "sort_column": {
      const { column_id, ascending, na_position } = transform;
      const args = argsList(
        `by="${column_id}"`,
        ascending ? "" : "ascending=False",
        na_position === "last" ? "" : `na_position="${na_position}"`,
      );
      return `${dfName}.sort_values(${args})`;
    }
    case "filter_rows": {
      const { operation, where } = transform;
      if (where.length === 0) {
        return dfName;
      }

      const whereClauses = where.map((condition) =>
        generateWhereClause(dfName, condition),
      );
      if (operation === "keep_rows" && whereClauses.length === 1) {
        return `${dfName}[${whereClauses[0]}]`;
      }

      const expression = whereClauses
        .map((clause) => `(${clause})`)
        .join(" & ");

      return operation === "keep_rows"
        ? `${dfName}[${expression}]`
        : `${dfName}[~(${expression})]`;
    }
    case "aggregate": {
      const { column_ids, aggregations } = transform;
      if (column_ids.length === 0) {
        return `${dfName}.agg(${listOfStrings(aggregations)})`;
      }
      return `${dfName}.agg({${column_ids
        .map((column_id) => `"${column_id}": ${listOfStrings(aggregations)}`)
        .join(", ")}})`;
    }
    case "group_by": {
      const { column_ids, aggregation, drop_na } = transform;
      const args = argsList(
        listOfStrings(column_ids),
        drop_na ? "dropna=True" : "",
      );
      return `${dfName}.groupby(${args}).${aggregation}()`;
    }
    case "select_columns": {
      const { column_ids } = transform;
      if (column_ids.length === 0) {
        return dfName;
      }
      if (column_ids.length === 1) {
        return `${dfName}["${column_ids[0]}"]`;
      }
      return `${dfName}[${listOfStrings(column_ids)}]`;
    }
    case "sample_rows": {
      const { n } = transform;
      return `${dfName}.sample(n=${n})`;
    }
    case "shuffle_rows":
      return `${dfName}.sample(frac=1)`;

    default:
      logNever(transform);
      return "";
  }
}

function generateWhereClause(
  dfName: string,
  where: {
    column_id: string;
    operator: OperatorType;
    value?: unknown;
  },
): string {
  const { column_id, operator, value } = where;
  switch (operator) {
    case "==":
      return `${dfName}["${column_id}"] == ${asString(value)}`;
    case "equals":
      return `${dfName}["${column_id}"].eq(${asString(value)})`;
    case "does_not_equal":
      return `${dfName}["${column_id}"].ne(${asString(value)})`;
    case "contains":
      return `${dfName}["${column_id}"].str.contains(${asString(value)})`;
    case "regex":
      return `${dfName}["${column_id}"].str.contains(${asString(
        value,
      )}, regex=True)`;
    case "starts_with":
      return `${dfName}["${column_id}"].str.startswith(${asString(value)})`;
    case "ends_with":
      return `${dfName}["${column_id}"].str.endswith(${asString(value)})`;
    case "in":
      return `${dfName}["${column_id}"].isin(${listOfStrings(value)})`;
    case "!=":
      return `${dfName}["${column_id}"].ne(${asString(value)})`;
    case ">":
      return `${dfName}["${column_id}"] > ${asString(value)}`;
    case ">=":
      return `${dfName}["${column_id}"] >= ${asString(value)}`;
    case "<":
      return `${dfName}["${column_id}"] < ${asString(value)}`;
    case "<=":
      return `${dfName}["${column_id}"] <= ${asString(value)}`;
    case "is_nan":
      return `${dfName}["${column_id}"].isna()`;
    case "is_not_nan":
      return `${dfName}["${column_id}"].notna()`;
    case "is_true":
      return `${dfName}["${column_id}"].eq(True)`;
    case "is_false":
      return `${dfName}["${column_id}"].eq(False)`;
    default:
      logNever(operator);
      return "df";
  }
}

function asString(value: unknown): string {
  if (typeof value === "string") {
    return `"${value}"`;
  }
  return `${value}`;
}

function listOfStrings(value: unknown): string {
  if (Array.isArray(value)) {
    return `[${value.map(asString).join(", ")}]`;
  }
  return asString(value);
}

function argsList(...args: string[]): string {
  return args.filter(Boolean).join(", ");
}
