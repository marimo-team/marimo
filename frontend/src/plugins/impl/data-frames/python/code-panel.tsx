/* Copyright 2024 Marimo. All rights reserved. */
import React from "react";
import { Transformations } from "../schema";
import { pythonPrintTransforms } from "./python-print";
import { ReadonlyPythonCode } from "@/components/editor/code/readonly-python-code";

interface Props {
  dataframeName: string;
  transforms?: Transformations;
}

export const CodePanel: React.FC<Props> = ({ transforms, dataframeName }) => {
  if (!transforms) {
    return null;
  }

  return (
    <ReadonlyPythonCode
      minHeight="215px"
      maxHeight="215px"
      code={pythonPrintTransforms(dataframeName, transforms.transforms)}
    />
  );
};
