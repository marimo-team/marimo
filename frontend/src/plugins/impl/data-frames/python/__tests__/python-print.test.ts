/* Copyright 2024 Marimo. All rights reserved. */
import { pythonPrint } from "@/plugins/impl/data-frames/python/python-print";
import { TransformType } from "@/plugins/impl/data-frames/schema";
import { expect, describe, it } from "vitest";
import {
  BOOLEAN_OPERATORS,
  DATE_OPERATORS,
  NUMERIC_OPERATORS,
  STRING_OPERATORS,
} from "../../utils/operators";
import { Objects } from "@/utils/objects";

describe("pythonPrint", () => {
  // Test for column_conversion
  it("generates correct Python code for column_conversion", () => {
    const transform: TransformType = {
      type: "column_conversion",
      column_id: "my_column",
      data_type: "int8",
      errors: "ignore",
    };
    const result = pythonPrint("df", transform);
    expect(result).toMatchInlineSnapshot(
      `"df["my_column"].astype("int8", errors="ignore")"`,
    );
  });

  // Test for rename_column
  it("generates correct Python code for rename_column", () => {
    const transform: TransformType = {
      type: "rename_column",
      column_id: "old_name",
      new_column_id: "new_name",
    };
    const result = pythonPrint("df", transform);
    expect(result).toMatchInlineSnapshot(
      `"df.rename(columns={"old_name": "new_name"})"`,
    );
  });

  // Test for sort_column
  it("generates correct Python code for sort_column", () => {
    const transform: TransformType = {
      type: "sort_column",
      column_id: "my_column",
      ascending: false,
      na_position: "first",
    };
    const result = pythonPrint("df", transform);
    expect(result).toMatchInlineSnapshot(
      `"df.sort_values(by="my_column", ascending=False, na_position="first")"`,
    );
  });

  // Test for aggregate
  it("generates correct Python code for aggregate", () => {
    const transform: TransformType = {
      type: "aggregate",
      column_ids: ["my_column"],
      aggregations: ["mean"],
    };
    const result = pythonPrint("df", transform);
    expect(result).toMatchInlineSnapshot(`"df.agg({"my_column": ["mean"]})"`);

    const transform2: TransformType = {
      type: "aggregate",
      column_ids: [],
      aggregations: ["mean"],
    };
    const result2 = pythonPrint("df", transform2);
    expect(result2).toMatchInlineSnapshot(`"df.agg(["mean"])"`);

    const transform3: TransformType = {
      type: "aggregate",
      column_ids: ["my_column"],
      aggregations: ["mean", "sum"],
    };
    const result3 = pythonPrint("df", transform3);
    expect(result3).toMatchInlineSnapshot(
      `"df.agg({"my_column": ["mean", "sum"]})"`,
    );
  });

  // Test for group_by
  it("generates correct Python code for group_by", () => {
    const transform: TransformType = {
      type: "group_by",
      column_ids: ["my_column"],
      aggregation: "sum",
      drop_na: true,
    };
    const result = pythonPrint("df", transform);
    expect(result).toMatchInlineSnapshot(
      `"df.groupby(["my_column"], dropna=True).sum()"`,
    );

    const transform2: TransformType = {
      type: "group_by",
      column_ids: ["my_column", "my_column2"],
      aggregation: "sum",
      drop_na: false,
    };
    const result2 = pythonPrint("df", transform2);
    expect(result2).toMatchInlineSnapshot(
      `"df.groupby(["my_column", "my_column2"]).sum()"`,
    );
  });
});

// Test for select_columns
it("generates correct Python code for select_columns", () => {
  const transform: TransformType = {
    type: "select_columns",
    column_ids: ["my_column"],
  };
  const result = pythonPrint("df", transform);
  expect(result).toMatchInlineSnapshot(`"df["my_column"]"`);

  const transform2: TransformType = {
    type: "select_columns",
    column_ids: ["my_column", "my_column2"],
  };
  const result2 = pythonPrint("df", transform2);
  expect(result2).toMatchInlineSnapshot(`"df[["my_column", "my_column2"]]"`);
});

// Test for sample_rows
it("generates correct Python code for sample_rows", () => {
  const transform: TransformType = {
    type: "sample_rows",
    n: 42,
    seed: 10,
    replace: false,
  };
  const result = pythonPrint("df", transform);
  expect(result).toMatchInlineSnapshot('"df.sample(n=42)"');
});

// Test for shuffle_rows
it("generates correct Python code for shuffle_rows", () => {
  const transform: TransformType = {
    type: "shuffle_rows",
    seed: 10,
  };
  const result = pythonPrint("df", transform);
  expect(result).toMatchInlineSnapshot('"df.sample(frac=1)"');
});

describe("pythonPrint: filter", () => {
  it("handles filter vs keep", () => {
    const result = pythonPrint("df", {
      type: "filter_rows",
      operation: "keep_rows",
      where: [
        {
          column_id: "my_column",
          operator: "==",
          value: 42,
        },
      ],
    });
    expect(result).toMatchInlineSnapshot(`"df[df["my_column"] == 42]"`);

    const result2 = pythonPrint("df", {
      type: "filter_rows",
      operation: "remove_rows",
      where: [
        {
          column_id: "my_column",
          operator: "==",
          value: 42,
        },
      ],
    });
    expect(result2).toMatchInlineSnapshot(`"df[~((df["my_column"] == 42))]"`);
  });

  it("handle multiple where clauses", () => {
    const result = pythonPrint("df", {
      type: "filter_rows",
      operation: "keep_rows",
      where: [
        {
          column_id: "my_column",
          operator: "==",
          value: 42,
        },
        {
          column_id: "my_column2",
          operator: "==",
          value: 43,
        },
      ],
    });
    expect(result).toMatchInlineSnapshot(
      `"df[(df["my_column"] == 42) & (df["my_column2"] == 43)]"`,
    );
  });

  // Test for filter_rows for strings
  it.each(Objects.entries(STRING_OPERATORS))(
    "filter_rows > string > %s",
    (operator) => {
      const transform: TransformType = {
        type: "filter_rows",
        operation: "keep_rows",
        where: [
          {
            column_id: "my_column",
            operator: operator,
            value: "val",
          },
        ],
      };
      const result = pythonPrint("df", transform);
      expect(result).toMatchSnapshot();
    },
  );

  // Test for filter_rows for booleans
  it.each(Objects.entries(BOOLEAN_OPERATORS))(
    "filter_rows > date > %s",
    (operator) => {
      const transform: TransformType = {
        type: "filter_rows",
        operation: "keep_rows",
        where: [
          {
            column_id: "my_column",
            operator: operator,
            value: true,
          },
        ],
      };
      const result = pythonPrint("df", transform);
      expect(result).toMatchSnapshot();
    },
  );

  // Test for filter_rows for dates
  it.each(Objects.entries(DATE_OPERATORS))(
    "filter_rows > date > %s",
    (operator) => {
      const transform: TransformType = {
        type: "filter_rows",
        operation: "keep_rows",
        where: [
          {
            column_id: "my_column",
            operator: operator,
            value: 1000,
          },
        ],
      };
      const result = pythonPrint("df", transform);
      expect(result).toMatchSnapshot();
    },
  );

  // Test for filter_rows for numbers
  it.each(Objects.entries(NUMERIC_OPERATORS))(
    "filter_rows > number > %s",
    (operator) => {
      const transform: TransformType = {
        type: "filter_rows",
        operation: "keep_rows",
        where: [
          {
            column_id: "my_column",
            operator: operator,
            value: 42,
          },
        ],
      };
      const result = pythonPrint("df", transform);
      expect(result).toMatchSnapshot();
    },
  );
});
