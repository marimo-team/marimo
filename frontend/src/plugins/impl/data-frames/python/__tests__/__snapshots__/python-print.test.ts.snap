// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pythonPrint: filter > filter_rows > date > != 1`] = `"df[df["my_column"].ne(1000)]"`;

exports[`pythonPrint: filter > filter_rows > date > < 1`] = `"df[df["my_column"] < 1000]"`;

exports[`pythonPrint: filter > filter_rows > date > <= 1`] = `"df[df["my_column"] <= 1000]"`;

exports[`pythonPrint: filter > filter_rows > date > == 1`] = `"df[df["my_column"] == 1000]"`;

exports[`pythonPrint: filter > filter_rows > date > > 1`] = `"df[df["my_column"] > 1000]"`;

exports[`pythonPrint: filter > filter_rows > date > >= 1`] = `"df[df["my_column"] >= 1000]"`;

exports[`pythonPrint: filter > filter_rows > date > is_false 1`] = `"df[df["my_column"].eq(False)]"`;

exports[`pythonPrint: filter > filter_rows > date > is_nan 1`] = `"df[df["my_column"].isna()]"`;

exports[`pythonPrint: filter > filter_rows > date > is_not_nan 1`] = `"df[df["my_column"].notna()]"`;

exports[`pythonPrint: filter > filter_rows > date > is_true 1`] = `"df[df["my_column"].eq(True)]"`;

exports[`pythonPrint: filter > filter_rows > number > != 1`] = `"df[df["my_column"].ne(42)]"`;

exports[`pythonPrint: filter > filter_rows > number > < 1`] = `"df[df["my_column"] < 42]"`;

exports[`pythonPrint: filter > filter_rows > number > <= 1`] = `"df[df["my_column"] <= 42]"`;

exports[`pythonPrint: filter > filter_rows > number > == 1`] = `"df[df["my_column"] == 42]"`;

exports[`pythonPrint: filter > filter_rows > number > > 1`] = `"df[df["my_column"] > 42]"`;

exports[`pythonPrint: filter > filter_rows > number > >= 1`] = `"df[df["my_column"] >= 42]"`;

exports[`pythonPrint: filter > filter_rows > number > is_nan 1`] = `"df[df["my_column"].isna()]"`;

exports[`pythonPrint: filter > filter_rows > number > is_not_nan 1`] = `"df[df["my_column"].notna()]"`;

exports[`pythonPrint: filter > filter_rows > string > contains 1`] = `"df[df["my_column"].str.contains("val")]"`;

exports[`pythonPrint: filter > filter_rows > string > does_not_equal 1`] = `"df[df["my_column"].ne("val")]"`;

exports[`pythonPrint: filter > filter_rows > string > ends_with 1`] = `"df[df["my_column"].str.endswith("val")]"`;

exports[`pythonPrint: filter > filter_rows > string > equals 1`] = `"df[df["my_column"].eq("val")]"`;

exports[`pythonPrint: filter > filter_rows > string > in 1`] = `"df[df["my_column"].isin("val")]"`;

exports[`pythonPrint: filter > filter_rows > string > regex 1`] = `"df[df["my_column"].str.contains("val", regex=True)]"`;

exports[`pythonPrint: filter > filter_rows > string > starts_with 1`] = `"df[df["my_column"].str.startswith("val")]"`;
